### ISSUE TITLE
Question: How to deal with a broken runtime?

### ISSUE THREAD

**Issue by bedeho** [Joystream/joystream#172]
# Definition

By _broken_ what is meant is that there is a bug somewhere in the implementation. By _deal with_, what is meant is what should you do in a module which has no static guarantee, but still expects some invariant to hold, which is only violated due to a broken runtime. Here are some examples

- **Unreserving funds:** You previously reserved some funds in the balance module, and now you are undeserving them.

- **Expecting a mapping:** You are trying to edit a post based on a `post_id`. In `edit(post_id,..)`, you also need to grab the thread to update some cumulative counter, e.g. `number_of_posts`, however, you are unable to find a `thread` in storage based on `post.thread_id`.

**In both cases, the only way the scenario can occur is if the module has a bug, or some other module you are using is not respecting the contract in its API, in which case it has a bug.**

# Parity Input

**Discussion with Shawn@Parity**
https://gist.github.com/bedeho/d5fa2080d3b4cbf8bf177f9b754adfaf

## TLDR

In the Substrate code base, the majority of cases where there _could_ be a broken runtime, such as call to unreserve funds, the assumption is made that this will not happen, and the possibility of an error is ignored, this is what Shawn refers to as _best effort_ in the chat above. He suggests as a policy that, the only place where one should depart from this is **where it may open you up to an attack due to someone avoiding getting charged for something**. So for example, the possible failure of trying to reserve or set aside funds is something where, even if it was not possible to fail in a correct runtime, he suggest you not ignore the error. 

# Approaches

There are at least three different ways of dealing with each such scenario.

## 1. Assert

This makes it very easy to understand what your system expects to hold to the reader, but if your system is actually broken, and you may find some hard to observe bugs while testing. We have indeed already caught bugs with asserts during testing. However, you also leave yourself open to a DoS attack on your chain, as no fees will be charged when the entire extrinsic is rolled back:

https://substrate.dev/docs/ru/tutorials/tcr/tcr-best-practices#handling-errors-gracefully

**Parity never does this.**

## 2. Ignore + debug asserts

By debug asserts, we mean protecting any assertions with something like

```Rust
#[cfg(debug_assertions)]
```

Beyond the asserts, the code flow proceeds as if the errors do not exist. This preserves the benefit of debugs as explained above, however, it avoids the issue of DoS, as they are not part of the production release.

**Parity does pure ignore in the majority of cases in its SRML codebase, no debug mode assertions.**

## 3. Model broken runtime errors

This means introducing an error case in your return types and possibly your own storage state.

### 3.1 Return types

Taking the case of  `fn edit_post(post_id) -> Result<S, E>`, it means adding one or more cases to type `E`, e.g.

```Rust
enum EditPostError{
BadPostId,
EditPostBrokenRuntime(EditPostBrokenRuntimeError)
}

enum EditPostBrokenRuntimeError {
BrokenRuntimeThreadMissing,
BrokenRuntimeFoo,
..
}
```

**Parity does this, but very rarely, e.g.**

https://github.com/paritytech/substrate/blob/d940c02440f80c5e0ded092c4dbc9f048312d8a6/frame/balances/src/lib.rs#L1028

### 3.2 In storage

Let's say that some extrinsic 

`bar() -> Result<(), BarError>`

in some other module tries to call 

`fn edit_post(post_id) -> Result<S, EditPostError>`

`bar` would now not only, as above, introduce som representation of `EditPostError` in `BarError`, but if this was detected _after mutations had begun_ then `bar` would store `EditPostBrokenRuntimeError` in its own storage representation. That would require introducing this into its own business logic types. e.g. if you had a map of post ids which were recently edited, you would have to do

```
RecentlyEditedPosts: map PostId => EditInfo

enum EditInfo {
Edited(...),
EditPostBrokenRuntimeError
}
```

**I have not seen Parity do this**

# What policy should we follow?

## A

- never do 1
- do 2
- never do 3.1
- never do 3.2

## B

- never do 1
- do 2
- do 3.1 on extremely rare occasions
- never do 3.2

## C

- never do 1
- do 2
- do 3.1 frequently 
- never do 3.2

## D

- never do 1
- do 2
- do 3.1 frequently 
- do 3.2 frequently

**Comment by bedeho** [comment]
We should have added the tentative conclusion here, anyone recall @mokhtar @shamil-gadelshin ?

**Comment by traumschule** [comment]
# opt-out

For the case that a runtime upgrade is expected (pending proposal, announce sudo action) and operators foresee bugs or want to perform a conservative fork for other reasons:

With the current system (a group of) operators have the option to modify the chain spec file and restart their node to disconnect from nodes that would accept the upgrade. This could be useful to keep some nodes as backup and optionally resync later to nodes that successfully upgraded, or the upgrade fails to deploy fixed code later (possible preemptive remedy for situations like the [Babylon chain split](https://github.com/Joystream/joystream/issues/2285)).
